site-src/geps/gep-1016.md

```diff
@@ -0,0 +1,813 @@
+}
+
+// GRPCRouteMatch defines the predicate used to match requests to a given
+// action. Multiple match types are ANDed together, i.e. the match will
+// evaluate to true only if all conditions are satisfied.
+//
+// For example, the match below will match a gRPC request only if its service
+// is `foo` AND it contains the `version: v1` header:
+//
+// ```
+// match:
+//   method:
+//     type: Exact
+//     service: "foo"
+//   headers:
+//   - name: "version"
+//     value "v1"
+// ```
+type GRPCRouteMatch struct {
+	// Path specifies a gRPC request service/method matcher. If this field is not
+	// specified, all services and methods will match.
+	//
+	// +optional
+	// +kubebuilder:default={type: "Exact"}
+	Method *GRPCMethodMatch `json:"path,omitempty"`
+
+	// Headers specifies gRPC request header matchers. Multiple match values are
+	// ANDed together, meaning, a request must match all the specified headers
+	// to select the route.
+	//
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Headers []GRPCHeaderMatch `json:"headers,omitempty"`
+}
+
+// GRPCPathMatch describes how to select a gRPC route by matching the gRPC
+// request service and/or method..
+type GRPCMethodMatch struct {
+	// Type specifies how to match against the service and/or method.
+	// Support: Core (Exact)
+	//
+	// Support: Custom (RegularExpression)
+	//
+	// +optional
+	// +kubebuilder:default=Exact
+	Type *GRPCMethodMatchType `json:"type,omitempty"`
+
+
+	// Value of the service to match against. If left empty or omitted, will
+	// match all services.
+	// +optional
+	// +kubebuilder:default=""
+	// +kubebuilder:validation:MaxLength=1024
+	Service *string `json:"value,omitempty"`
+
+	// Value of the method to match against. If left empty or omitted, will
+	// match all services.
+	// +optional
+	// +kubebuilder:default=""
+	// +kubebuilder:validation:MaxLength=1024
+	Method *string `json:"value,omitempty"`
+}
+
+// MethodMatchType specifies the semantics of how gRPC methods and services should be compared.
+// Valid MethodMatchType values are:
+//
+// * "Exact"
+// * "RegularExpression"
+//
+// Exact paths must be syntactically valid:
+//
+// - Must not contain `/` character
+//
+// +kubebuilder:validation:Enum=Exact;PathPrefix;RegularExpression
+// +kubebuilder:validation:Enum=Exact;RegularExpression
+type GRPCMethodMatchType string
+
+const (
+	// Matches the service and/or method exactly and with case sensitivity.
+	PathMatchExact PathMatchType = "Exact"
+
+	// Matches if the service and/or method matches the given regular expression with
+	// case sensitivity.
+	//
+	// Since `"RegularExpression"` has custom conformance, implementations
+	// can support POSIX, PCRE, RE2 or any other regular expression dialect.
+	// Please read the implementation's documentation to determine the supported
+	// dialect.
+	PathMatchRegularExpression PathMatchType = "RegularExpression"
```

Comment:
Would it make sense to also consider matching based on the prefix ? For example, to match based on the package:
```yaml
  rules:
    matches:
    - service: foo. # route traffic that belongs to package foo
      type: Prefix
    backendRefs:
    - name: grpc-backend
      port: 50051
 ```
 
It looks like it is possible to configure that using the RegularExpression match type. However, prefix matching in proxies are more efficient than regex-based matching.

Is it a viable use case?

Reply:
I've considered this. The problem is that allowing prefix matching on the _service_ with a method specified would actually be an HTTP/2 _infix_ matcher. For example, consider the following:

```
  rules:
    matches:
    - service: foo.
      method: bar
      type: Prefix
```

This would match HTTP/2 URIs conforming to this regex:

```
/foo\..*/bar.*
```

It also becomes much more difficult to justify the fact that both the service and the method share a matcher type if you allow prefix match. Even if you had both a `serviceType` and a `methodType`, you still have this infix matcher problem though.

My preference is to keep it simple for the moment and go with what's currently included in the proposal. If folks feel strongly about this though, I'm happy to hear the feedback.